pkg/reconciler/finalizer/reposync_finalizer_test.go

```diff
@@ -271,11 +278,181 @@ func TestRepoSyncFinalize(t *testing.T) {
+				}
+				return obj
+			}(),
+			expectedRsyncBeforeDestroy: func() client.Object {
+				obj := repoSync1.DeepCopy()
+				// No changes - continue deleting
+				obj.SetResourceVersion("3")
+				obj.Status.Conditions = []v1beta1.RepoSyncCondition{
+					{
+						Type:    v1beta1.RepoSyncReconcilerFinalizing,
+						Status:  metav1.ConditionTrue,
+						Reason:  "ResourcesDeleting",
+						Message: "Deleting managed resource objects",
+					},
+					{
+						Type:    v1beta1.RepoSyncReconcilerFinalizerFailure,
+						Status:  metav1.ConditionTrue,
+						Reason:  "DestroyFailure",
+						Message: "Failed to delete managed resource objects",
+						Errors: []v1beta1.ConfigSyncError{
+							{
+								Code:         "2002",
+								ErrorMessage: "KNV2002: example message: APIServer error: destroy error\n\nFor more information, see https://g.co/cloud/acm-errors#knv2002",
+							},
+						},
+					},
+				}
+				return obj
+			}(),
+			destroyErrs:     nil,
+			expectedError:   nil,
+			expectedStopped: true,
+			expectedRsyncAfterFinalize: func() client.Object {
+				obj := repoSync1.DeepCopy()
+				// +1 to remove ReconcilerFinalizing condition
+				// +1 to remove ReconcilerFinalizerFailure condition
+				// +1 to remove Finalizer
+				// TODO: optimize by combining consecutive updates
+				obj.SetResourceVersion("6")
+				// Finalizer has been removed
+				obj.SetFinalizers(nil)
+				// ReconcilerFinalizing condition removed
+				// ReconcilerFinalizerFailure condition removed
+				return obj
+			}(),
+		},
 	}
 
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
-			fakeClient := fake.NewClient(t, scheme, tc.rsync)
+			// Set up a toy object referenced by the ResourceGroup
+			cm := k8sobjects.ConfigMapObject(
+				core.Name("cm1"), core.Namespace(tc.rsync.GetNamespace()),
+				core.Annotation("a-key", "a-value"),
+				core.Label("l-key", "l-value"))
+			wantCM := cm.DeepCopy()
+			wantCM.Generation = 1
+			wantCM.ResourceVersion = "1"
+			wantCM.UID = "1"
+
+			applySetID := "apply-set-id"
+			if tc.deletionPolicy == metadata.DeletionPropagationPolicyForeground {
+				wantCM.SetAnnotations(map[string]string{
+					metadata.OwningInventoryKey:          "inventory-id",
+					metadata.ManagementModeAnnotationKey: metadata.ManagementEnabled.String(),
+					metadata.SyncTokenAnnotationKey:      "commit-hash",
+					metadata.GitContextKey:               "git-context",
+					metadata.ResourceManagerKey:          "manager-value",
+					metadata.ResourceIDKey:               "_configmap_example_cm1",
+					"a-key":                              "a-value",
+				})
+				wantCM.SetLabels(map[string]string{
+					metadata.ManagedByKey:        metadata.ManagedByValue,
+					metadata.ApplySetPartOfLabel: applySetID,
+					"l-key":                      "l-value",
+				})
+			} else { // When "Orphan" is used, all CS metadata should be removed from the CM
+				wantCM.ResourceVersion = "2"
+				wantCM.SetAnnotations(map[string]string{
```

Comment:
Looks like these can be pulled out and put up above the if statement. Then you can call SetConfigSyncMetadata on the wantCM if the policy is foreground.

Reply:
It was intentional to not use `SetConfigSyncMetadata` to compute the expected value. I want the assertion to be explicit/unambiguous and not rely on the business logic to test the business logic.

Relevant/recent [tott](https://goto.google.com/tott/741)